backend-rust/src/graphql_api.rs

```diff
@@ -905,14 +905,92 @@ impl SearchResult {
 
     async fn blocks(
         &self,
-        #[graphql(desc = "Returns the first _n_ elements from the list.")] _first: Option<i32>,
+        ctx: &Context<'_>,
+        #[graphql(desc = "Returns the first _n_ elements from the list.")] first: Option<u64>,
         #[graphql(desc = "Returns the elements in the list that come after the specified cursor.")]
-        _after: Option<String>,
-        #[graphql(desc = "Returns the last _n_ elements from the list.")] _last: Option<i32>,
+        after: Option<String>,
+        #[graphql(desc = "Returns the last _n_ elements from the list.")] last: Option<u64>,
         #[graphql(desc = "Returns the elements in the list that come before the specified cursor.")]
-        _before: Option<String>,
+        before: Option<String>,
     ) -> ApiResult<connection::Connection<String, Block>> {
-        todo_api!()
+        let block_hash_regex: Regex = Regex::new(r"^[a-fA-F0-9]{1,64}$")
+            .map_err(|_| ApiError::InternalError("Invalid regex".to_string()))?;
+        let pool = get_pool(ctx)?;
+        let config = get_config(ctx)?;
+        let query =
+            ConnectionQuery::<i64>::new(first, after, last, before, config.block_connection_limit)?;
+        let mut connection = connection::Connection::new(false, false);
+        if !block_hash_regex.is_match(&self.query) {
+            return Ok(connection);
+        }
+        let lower_case_query = self.query.to_lowercase();
+        let rows = sqlx::query_as!(
+            Block,
+            "SELECT * FROM (
+                SELECT
+                    hash,
+                    height,
+                    slot_time,
+                    block_time,
+                    finalization_time,
+                    baker_id,
+                    total_amount
+                FROM blocks
+                WHERE
+                    height = $5
+                    OR starts_with(hash, $6)
+                    AND height > $1
+                    AND height < $2
+                ORDER BY
+                    (CASE WHEN $4 THEN height END) ASC,
+                    (CASE WHEN NOT $4 THEN height END) DESC
+                LIMIT $3
+            ) ORDER BY height DESC",
+            query.from,
+            query.to,
+            query.limit,
+            query.desc,
+            lower_case_query.parse::<i64>().ok(),
+            lower_case_query
+        )
+        .fetch_all(pool)
```

Comment:
The advantage of `fetch` over `fetch_all` is that we can avoid allocating the Vec first.
Minimal optimization, but results in the same amount of code, so why not.